File: src/main/java/gargoyle/ct/CTTimeUtil.java
package gargoyle.ct;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.concurrent.TimeUnit;

public final class CTTimeUtil {

    private static final String HH_MM_SS = "HH:mm:ss";

    private static final String MM = "mm";

    private static final String MM_SS = "mm:ss";

    private static final String SS = "ss";

    private CTTimeUtil() {
    }

    public static long convert(TimeUnit unitTo, TimeUnit unitFrom, long cnt) {
        return unitTo.convert(unitFrom.toMillis(cnt), TimeUnit.MILLISECONDS);
    }

    public static long currentTimeMillis() {
        return new Date().getTime();
    }

    public static long downTo(long currentMillis, long baseMillis) {
        return currentMillis / baseMillis * baseMillis;
    }

    private static String format(String format, long currentMillis) {
        return new SimpleDateFormat(format, Locale.getDefault()).format(new Date(currentMillis));
    }

    public static String formatHHMMSS(long currentMillis) {
        return format(HH_MM_SS, currentMillis);
    }

    public static String formatMM(long currentMillis) {
        return format(MM, currentMillis);
    }

    public static String formatMMSS(long currentMillis) {
        return format(MM_SS, currentMillis);
    }

    public static String formatSS(long currentMillis) {
        return format(SS, currentMillis);
    }

    public static long fromMillis(TimeUnit unit, long millis) {
        return unit.convert(millis, TimeUnit.MILLISECONDS);
    }

    public static boolean isBetween(long currentMillis, long startMillis, long endMillis) {
        return currentMillis >= startMillis && currentMillis <= endMillis;
    }

    public static boolean isBetween(TimeUnit unit, long current, long start, long end) {
        return toMillis(unit, current) >= toMillis(unit, start) && toMillis(unit, current) <= toMillis(unit, end);
    }

    public static boolean isInPeriod(TimeUnit unit, long currentMillis, int period, int delay) {
        return fromMillis(unit, currentMillis) % period < delay;
    }

    public static long make() {
        Calendar calendar = Calendar.getInstance();
        calendar.set(Calendar.MILLISECOND, 0);
        return calendar.getTimeInMillis();
    }

    public static long make(int hours, int minutes, int seconds) {
        Calendar calendar = Calendar.getInstance();
        calendar.set(Calendar.HOUR_OF_DAY, hours);
        calendar.set(Calendar.MINUTE, minutes);
        calendar.set(Calendar.SECOND, seconds);
        calendar.set(Calendar.MILLISECOND, 0);
        return calendar.getTimeInMillis();
    }

    public static long parseHHMMSS(String string) {
        String[] pair = string.split(":");
        return make(Integer.parseInt(pair[0]), Integer.parseInt(pair[1]), Integer.parseInt(pair[2]));
    }

    private static long timeElapsedFrom(long currentMillis, long begin) {
        return currentMillis - begin;
    }

    public static long timeElapsedFrom(TimeUnit unit, long currentMillis, long begin) {
        return fromMillis(unit, timeElapsedFrom(currentMillis, begin));
    }

    public static long timeRemainsTo(long currentMillis, long end) {
        return end - currentMillis;
    }

    public static long timeRemainsTo(TimeUnit unit, long currentMillis, long end) {
        return fromMillis(unit, timeRemainsTo(currentMillis, end));
    }

    public static long toBase(long startMillis, long currentMillis, long baseMillis) {
        long low = downTo(currentMillis, baseMillis);
        long high = upTo(currentMillis, baseMillis);
        long ret = startMillis;
        while (ret <= low) {
            ret += baseMillis;
        }
        while (ret >= high) {
            ret -= baseMillis;
        }
        return ret;
    }

    public static long toMillis(TimeUnit unit, long duration) {
        return unit.toMillis(duration);
    }

    public static long upTo(long currentMillis, long baseMillis) {
        return downTo(currentMillis, baseMillis) + baseMillis;
    }
}

File: util/src/main/kotlin/gargoyle/ct/util/util/CTTimeUtil.kt
package gargoyle.ct.util.util

import java.time.*
import java.time.format.DateTimeFormatter
import java.util.*
import java.util.concurrent.TimeUnit

object CTTimeUtil {
    private const val HH_MM_SS = "HH:mm:ss"
    private const val MM = "mm"
    private const val MM_SS = "mm:ss"
    private const val SS = "ss"

    fun convert(unitTo: TimeUnit, unitFrom: TimeUnit, cnt: Long): Long =
        unitTo.convert(unitFrom.toMillis(cnt), TimeUnit.MILLISECONDS)

    fun currentTimeMillis(): Long = Date().time

    fun formatHHMMSS(currentMillis: Long): String = format(HH_MM_SS, currentMillis)

    private fun format(format: String, currentMillis: Long): String =
        Instant.ofEpochMilli(currentMillis).atOffset(ZoneOffset.UTC).format(DateTimeFormatter.ofPattern(format))

    fun formatMM(currentMillis: Long): String = format(MM, currentMillis)

    fun formatMMSS(currentMillis: Long): String = format(MM_SS, currentMillis)

    fun formatSS(currentMillis: Long): String = format(SS, currentMillis)

    fun isBetween(currentMillis: Long, startMillis: Long, endMillis: Long): Boolean =
        currentMillis in startMillis..endMillis

    fun isBetween(unit: TimeUnit, current: Long, start: Long, end: Long): Boolean =
        toMillis(unit, current) >= toMillis(unit, start) && toMillis(unit, current) <= toMillis(unit, end)

    fun toMillis(unit: TimeUnit, duration: Long): Long = unit.toMillis(duration)

    fun isInPeriod(unit: TimeUnit, currentMillis: Long, period: Int, delay: Int): Boolean =
        fromMillis(unit, currentMillis) % period < delay

    fun fromMillis(unit: TimeUnit, millis: Long): Long = unit.convert(millis, TimeUnit.MILLISECONDS)

    fun make(): Long = LocalDateTime.now().withNano(0).toInstant(ZoneOffset.UTC).toEpochMilli()

    fun parseHHMMSS(string: String): Long {
        val pair = string.split(':').toTypedArray()
        return make(pair[0].toInt(), pair[1].toInt(), pair[2].toInt())
    }

    fun make(hours: Int, minutes: Int, seconds: Int): Long {
        return LocalDateTime.of(LocalDate.now(), LocalTime.of(hours, minutes, seconds))
            .withNano(0)
            .toInstant(ZoneOffset.UTC)
            .toEpochMilli()
    }

    fun timeElapsedFrom(unit: TimeUnit, currentMillis: Long, begin: Long): Long =
        fromMillis(unit, timeElapsedFrom(currentMillis, begin))

    private fun timeElapsedFrom(currentMillis: Long, begin: Long): Long = currentMillis - begin

    fun timeRemainsTo(unit: TimeUnit, currentMillis: Long, end: Long): Long =
        fromMillis(unit, timeRemainsTo(currentMillis, end))

    fun timeRemainsTo(currentMillis: Long, end: Long): Long = end - currentMillis

    fun toBase(startMillis: Long, currentMillis: Long, baseMillis: Long): Long {
        val low = downTo(currentMillis, baseMillis)
        val high = upTo(currentMillis, baseMillis)
        var ret = startMillis
        while (ret <= low) {
            ret += baseMillis
        }
        while (ret >= high) {
            ret -= baseMillis
        }
        return ret
    }

    fun downTo(currentMillis: Long, baseMillis: Long): Long = currentMillis / baseMillis * baseMillis

    fun upTo(currentMillis: Long, baseMillis: Long): Long = downTo(currentMillis, baseMillis) + baseMillis
}
